DAWN/examples/gpt_lightning/README.md

```diff
@@ -0,0 +1,478 @@
+export USE_XETLA=OFF
+
+# Use immediate command lists for performance
+# See https://www.intel.com/content/www/us/en/developer/articles/guide/level-zero-immediate-command-lists.html
+export SYCL_PI_LEVEL_ZERO_USE_IMMEDIATE_COMMANDLISTS=1
+
+# Define the master communication node for MPI
+# See https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.plugins.environments.SLURMEnvironment.html
+export MASTER_ADDR=${SLURMD_NODENAME}
+# Select a random port in the range [29880, 38072)
+export MASTER_PORT=$((29880 + ($RANDOM % 8192)))
+```
+
+The last two environment variables are `MASTER_ADDR` and `MASTER_PORT`. These are the name and port for the master node that’s used to support communication between all of the other nodes. We’re get the name of the master node from SLURM and choose a random port.
+
+### Setting up the build environment
+
+The OneAPI documentation tells us that we need to run some scripts to set up the environment. Having activated the various modules we still need to do this and the following commands are essentially just following these requirements.
+
+```shell
+echo "Initialise environment"
+pushd /usr/local/dawn/software/spack/spack-views/dawn-test-2023-12-22/
+source intel-oneapi-compilers-2024.0.0/gcc-13.2.0/znjudqsiaf6x5u2rxdtymf6ss55nmimw/compiler/2024.0/env/vars.sh
+source intel-oneapi-mkl-2024.0.0/oneapi-2024.0.0/4n7ruz44nhbsd5xp4nnz6mgm2z7vqzxs/mkl/2024.0/env/vars.sh
+source intel-oneapi-compilers-2024.0.0/gcc-13.2.0/znjudqsiaf6x5u2rxdtymf6ss55nmimw/setvars.sh
+popd
+```
+
+### Setting up the Virtual Environment
+
+This next section is a bit of a mess, but it’s actually quite simple. The commands check whether the Python virtual environment directory already exists. If it doesn’t it creates the virtual environment and installed all of the requirements inside it, based on the list of packages in the various `pip` requirements files. If it does exist it assumes everything is already correctly installed and just activates the environment instead. 
+
+```shell
+if [ -d "$VENV_PATH" ]; then
+  echo "Activate virtual environment"
+  source "${VENV_PATH}"/bin/activate
+  source $(python -c "import oneccl_bindings_for_pytorch as torch_ccl;print(torch_ccl.cwd)")/env/vars.sh
+else
+  echo "Create virtual environment"
+  python3.9 -m venv ${VENV_PATH}
+  source ./${VENV_PATH}/bin/activate
+  pip install --upgrade pip
+  pip install torch==2.0.1a0 torchvision==0.15.2a0 intel-extension-for-pytorch==2.0.120+xpu oneccl-bind-pt==2.0.200 --extra-index-url https://pytorch-extension.intel.com/release-whl-aitools/
+  source $(python -c "import oneccl_bindings_for_pytorch as torch_ccl;print(torch_ccl.cwd)")/env/vars.sh
+  pip install -e .
+  pip install -r requirements.txt
+  pip install -r requirements/nanogpt.txt
+fi
+```
+
+This means that the first time we run the batch file a suitable Python virtual environment will be created. On subsequent runs it will activate the virtual environment instead.
+
```

Comment:
This information may be better moved since it applies to all examples. I would recommend creating a separate MD file with tips and tricks.

Reply:
Thanks for suggesting this; it's a nice idea. I've created a new 'tips-and-tricks' folder and moved this content into a file that covers SLURM batch files more generally. I agree, this is a much nicer way to structure things.

Reply:
Thanks for suggesting this; it's a nice idea. I've created a new 'tips-and-tricks' folder and moved this content into a file that covers SLURM batch files more generally. I agree, this is a much nicer way to structure things.